packages/formspree-core/src/submission.ts

```diff
@@ -0,0 +1,181 @@
+  obj: object
+): obj is ServerStripePluginPendingResponse {
+  if (
+    'stripe' in obj &&
+    'resubmitKey' in obj &&
+    typeof obj.resubmitKey === 'string'
+  ) {
+    const { stripe } = obj;
+    return (
+      typeof stripe === 'object' &&
+      stripe != null &&
+      'paymentIntentClientSecret' in stripe &&
+      typeof stripe.paymentIntentClientSecret === 'string'
+    );
+  }
+  return false;
+}
+
+export class SubmissionErrorResult<T extends FieldValues> {
+  readonly kind = 'error';
+
+  private readonly formErrors: FormError[] = [];
+  private readonly fieldErrors: Map<keyof T, FieldError[]> = new Map();
+
+  constructor(...serverErrors: ServerError[]) {
+    for (const err of serverErrors) {
+      // form errors
+      if (!err.field) {
+        this.formErrors.push({
+          code:
+            err.code && isFormErrorCode(err.code) ? err.code : 'UNSPECIFIED',
+          message: err.message,
+        });
+        continue;
+      }
+
+      const fieldErrors = this.fieldErrors.get(err.field) ?? [];
+      fieldErrors.push({
+        code: err.code && isFieldErrorCode(err.code) ? err.code : 'UNSPECIFIED',
+        message: err.message,
+      });
+      this.fieldErrors.set(err.field, fieldErrors);
+    }
+  }
+
+  getFormErrors(): readonly FormError[] {
+    return [...this.formErrors];
+  }
+
+  getFieldErrors<K extends keyof T>(field: K): readonly FieldError[] {
+    return this.fieldErrors.get(field) ?? [];
+  }
+
+  getAllFieldErrors(): readonly [keyof T, readonly FieldError[]][] {
+    return Array.from(this.fieldErrors);
+  }
+}
+
+export const emptySubmissionErrorResult = new SubmissionErrorResult();
+
+export type FormError = {
+  readonly code: FormErrorCode | 'UNSPECIFIED';
+  readonly message: string;
+};
+
+function isFormErrorCode(code: string): code is FormErrorCode {
+  return code in FormErrorCodeEnum;
+}
+
+export type FormErrorCode = ValueOf<typeof FormErrorCodeEnum>;
+
+export const FormErrorCodeEnum = {
+  BLOCKED: 'BLOCKED',
+  EMPTY: 'EMPTY',
+  FILES_TOO_BIG: 'FILES_TOO_BIG',
+  FORM_NOT_FOUND: 'FORM_NOT_FOUND',
+  INACTIVE: 'INACTIVE',
+  NO_FILE_UPLOADS: 'NO_FILE_UPLOADS',
+  PROJECT_NOT_FOUND: 'PROJECT_NOT_FOUND',
+  TOO_MANY_FILES: 'TOO_MANY_FILES',
+  // Stripe
+  STRIPE_CLIENT_ERROR: 'STRIPE_CLIENT_ERROR',
+  STRIPE_SCA_ERROR: 'STRIPE_SCA_ERROR',
+} as const;
+
+export type FieldError = {
+  readonly code: FieldErrorCode | 'UNSPECIFIED';
+  readonly message: string;
+};
+
+function isFieldErrorCode(code: string): code is FieldErrorCode {
+  return code in FieldErrorCodeEnum;
+}
+
+export type FieldErrorCode = ValueOf<typeof FieldErrorCodeEnum>;
+
+export const FieldErrorCodeEnum = {
+  REQUIRED_FIELD_EMPTY: 'REQUIRED_FIELD_EMPTY',
+  REQUIRED_FIELD_MISSING: 'REQUIRED_FIELD_MISSING',
+  TYPE_EMAIL: 'TYPE_EMAIL',
+  TYPE_NUMERIC: 'TYPE_NUMERIC',
+  TYPE_TEXT: 'TYPE_TEXT',
+  // Stripe
+  STRIPE_CLIENT_ERROR: 'STRIPE_CLIENT_ERROR',
+  STRIPE_SCA_ERROR: 'STRIPE_SCA_ERROR',
```

Comment:
@colevscode are these field errors or form errors? I think they are form errors, however, the server sends the error payload back tying it to the `paymentMethod` field. If they should be form errors, we'll need to special case these in https://github.com/formspree/formspree-js/pull/47/files#diff-1f9db01dd19d42e313a63f565d5ea2998a592f709e20fa1ec56f4a816f57ebdaR77-R96.

Reply:
> The idea is that the user could place stripe error messages near the stripe card element with <ValidationError field='paymentMethod'> even though they don't have a paymentMethod named input.

Perfect. Then it'll just work.

`FormError` and `FieldError` are not designed to be a public interface since we provide a `SubmissionError` result type and users will get form/field errors from its method.

We should export `FieldErrorCode` and `FormErrorCode` though. Since we define these error codes, they are part of the public interface.